### DISCUSSION TITLE
Text Bounding Box Below Actual Text Position

### DISCUSSION THREAD

**Discussion by renraeldab** [pymupdf/PyMuPDF#3772] [Looking for help] [answered]
I guess this can happen when there are some uncommon fonts. Is there any way to detect or fix this?
![image](https://github.com/user-attachments/assets/cdd29530-0a06-43fe-8246-cf6b6246c89b)

**Comment by JorjMcKie** [comment]
No, this cannot be detected: we are confined to believe what a font is telling us ...
It **_may_** be a bug though - the picture does look a bit weird. It might be worthwhile to take a look at the page.
You could marks the rectangles based on extraction data after you have executed `pymupdf.TOOLS.set_small_glyph_heights(True)`. This option causes PyMuPDF's character bbox computations to take a different path.
If this phenomenon then goes away, we do have a bug (in MuPDF).

**Reply by renraeldab** [reply]
Thanks for your quick response. I already added  pymupdf.TOOLS.set_small_glyph_heights(True) after I found this, but it didn't fix those wrong bounding boxes. Then I thought the pdf file might cause this problem. However, I ran a quick test on pdfminer, and it seemed normal.

The result of pdfminer.high_level.extract_pages:
![image](https://github.com/user-attachments/assets/4b86fb71-6a9b-40ea-9896-1eff05f7dba5)

**Reply by JorjMcKie** [reply]
Ok, thanks.
Please let us have the PDF page. I would like to have a look at it.

**Reply by renraeldab** [reply]
Link to the pdf: https://iiis.tsinghua.edu.cn/index.php?f=down&v=d&s=T4HPwJkpEXiUvs1bOwgbD6x2Md7WKMrrZWjYQHDPNOvrZhh6RmhU5S007nqjFuXriXO72831ol7QbJlWKuzehqwSIDcdN%2BwH

**Reply by JorjMcKie** [reply]
Thanks for the file.
I looked at the used fonts and found one that is erroneous in that it contains crazy values for the so-called ascender and descender values. Please look at the definition of these properties [here](https://pymupdf.readthedocs.io/en/latest/textpage.html#span-dictionary).
Both are crucial for computing text positions. Ascender must be positive and descender must be negative. This is what we find:
```python
for itm in page.get_fonts():
    buff=doc.extract_font(itm[0])[-1]
    font=pymupdf.Font(fontbuffer=buff)
    print(f"xref {itm[0]}, {font.name=}, {font.ascender=}, {font.descender=}")

    
xref 144, font.name='AdvOT735b620e Regular', font.ascender=1.0, font.descender=-0.29100000858306885
xref 145, font.name='AdvOT144a03c1 Regular', font.ascender=-32.768001556396484, font.descender=-0.25999999046325684
xref 146, font.name='AdvOT5e4d79fc Regular', font.ascender=0.8539999723434448, font.descender=-0.25
xref 155, font.name='AdvOT1ef757c0 Regular', font.ascender=0.8539999723434448, font.descender=-0.25
xref 165, font.name='AdvTT16f3b945.B Regular', font.ascender=0.9580000042915344, font.descender=-0.25
xref 166, font.name='AdvOT1415ea69.B Regular', font.ascender=1.0, font.descender=-0.25999999046325684
xref 167, font.name='AdvOTf32ed098.B Regular', font.ascender=1.0099999904632568, font.descender=-0.25999999046325684
```
Font at xref 145 shows an ascender value of ridiculous -32.768. Unacceptable for position computations: some reasonable values must be used instead.
Unfortunately, this corrective algorithm does not work satisfactorily here ...

**Reply by renraeldab** [reply]
The ascenders and descenders are all normal in get_text("dict") results, different from the fonts info. Is this after correction?

**Reply by JorjMcKie** [reply]
Yes - but the subsequent code does not correctly used the corrected ascender / descender values.
This is in contrast to `page.get_texttrace()` - a method acting independently of a `TextPage`. The per-character boundary boxes here are sufficiently corrected.

**Comment by JorjMcKie** [comment, marked as answer]
I think that MuPDF makes an error here. After confirmation, I will submit a bug into their system.
